proof/refine/AARCH64/ArchStateRelationLemmas.thy

```diff
@@ -0,0 +1,176 @@
+(*
+ * Copyright 2023, Proofcraft Pty Ltd
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
+ * SPDX-License-Identifier: GPL-2.0-only
+ *)
+
+(*
+   Arch-specific lemmas related to the refinement relation between abstract and concrete states
+*)
+
+theory ArchStateRelationLemmas
+imports StateRelation
+begin
+
+context Arch begin arch_global_naming
+
+named_theorems StateRelation_R_assms
+
+lemma obj_relation_cuts_def2:
+  "obj_relation_cuts ko x =
+   (case ko of CNode sz cs \<Rightarrow> if well_formed_cnode_n sz cs
+                              then {(cte_map (x, y), cte_relation y) | y. y \<in> dom cs}
+                              else {(x, \<bottom>\<bottom>)}
+             | TCB tcb \<Rightarrow> {(x, tcb_relation_cut)}
+             | ArchObj (PageTable pt) \<Rightarrow> (\<lambda>y. (x + (y << pteBits), pte_relation y)) `
+                                         {0..mask (ptTranslationBits (pt_type pt))}
+             | ArchObj (DataPage dev sz) \<Rightarrow>
+                 {(x + (n << pageBits),  \<lambda>_ obj. obj =(if dev then KOUserDataDevice else KOUserData))
+                  | n . n < 2 ^ (pageBitsForSize sz - pageBits) }
+             | ArchObj _ \<Rightarrow> {(x, other_aobj_relation)}
+             | _ \<Rightarrow> {(x, other_obj_relation)})"
+  by (simp split: Structures_A.kernel_object.split
+                  AARCH64_A.arch_kernel_obj.split)
+
+lemma obj_relation_cuts_def3:
+  "obj_relation_cuts ko x =
+   (case a_type ko of
+      ACapTable n \<Rightarrow> {(cte_map (x, y), cte_relation y) | y. length y = n}
+    | ATCB \<Rightarrow> {(x, tcb_relation_cut)}
+    | AArch (APageTable pt_t) \<Rightarrow> (\<lambda>y. (x + (y << pteBits), pte_relation y)) `
+                                      {0..mask (ptTranslationBits pt_t)}
+    | AArch (AUserData sz) \<Rightarrow> {(x + (n << pageBits), \<lambda>_ obj. obj = KOUserData)
+                               | n . n < 2 ^ (pageBitsForSize sz - pageBits) }
+    | AArch (ADeviceData sz) \<Rightarrow> {(x + (n << pageBits), \<lambda>_ obj. obj = KOUserDataDevice )
+                                 | n . n < 2 ^ (pageBitsForSize sz - pageBits) }
+    | AArch _ \<Rightarrow> {(x, other_aobj_relation)}
+    | AGarbage _ \<Rightarrow> {(x, \<bottom>\<bottom>)}
+    | _ \<Rightarrow> {(x, other_obj_relation)})"
+  by (simp add: obj_relation_cuts_def2 a_type_def well_formed_cnode_n_def length_set_helper
+           split: Structures_A.kernel_object.split AARCH64_A.arch_kernel_obj.split)
+
+(* FIXME arch-split: split off generic? *)
+lemma obj_relation_cutsE:
```

Comment:
This is one of the more fundamental questions. If we have case distinction lemmas like this in here, that means that all lemmas that use them must be arch specific. That's not necessarily a problem, but it is a potential point for optimisation later.

I'd vote for leaving the FIXME in for now and see how it goes with just the arch-specific version of it. If it draws in a lot of later lemmas that we really would like to be generic, then we should come back here and split it up into generic cases and arch cases. Doing it immediately feels like premature optimisation.